kip-0015.md

```diff
@@ -0,0 +1,304 @@
+---
+KIP: 0015
+Title: quicksign signing api v1
+Author: Jacquin Mininger @jmininger, Doug Beardsley @mightybyte
+Status: Draft
+Type: Standard
+Category: Interface
+Created: 2022-09-15
+---
+
+# Abstract
+
+We propose an additional `quicksign` endpoint to the [kadena-signing-api](https://github.com/kadena-io/signing-api) that allows a wallet to show the user multiple transactions at once for signature approval
+
+# Motivation
+
+The current `sign` method used by `signing-api-v1` contains a number of limitations:
+
+- Only a single txn can be signed at a time. This is painfully inefficient for workflows that require multi-step transactions or for transactions to occur across multiple chains
+- Only `exec` transactions can be sent and signed; `cont` transactions have no means of being signed in the current api
+- The response value returns a `Command` type which represents a fully-signed transaction. It does not allow for an intermediate state with a partially signed transaction. The latter is a useful workflow given that pact is multisig-native
+
+In addition to the above, we propose an evolution of the `SigData` type used by the pact command line tool for representing partially-signed payloads (since pact’s `Command` type can only exist as a fully signed payload), that we call `CommandSigData`.
+
+- The `SigData` type includes a `hash` field [https://github.com/kadena-io/pact/blob/957b8bd7644cebc60d043b33f0d7ffc53c65d783/src/Pact/Types/SigData.hs#L62-L66](https://github.com/kadena-io/pact/blob/957b8bd7644cebc60d043b33f0d7ffc53c65d783/src/Pact/Types/SigData.hs#L62-L66) . This data can also be derived from the `cmd` field of `SigData`. Imagine a scenario where a malicious dapp sends a `SigData` where the hash represents a malicious transaction, while the payload represents an unrelated transaction that the user is expecting. If the wallet does not explicitly check the hash, by recalculating, this becomes a security issue. To avoid this we force the wallet to compute the hash (by omitting it from the new `CommandSigData` structure)
+
+# Specification
+
+## Schema Sketch
+
+### Request
+```
+{
+  "commandSigDatas": [ { "cmd": string, "signatureList": [ { "publicKey":string, "signature": string|null }]}]
```

Comment:
@jmininger After discussing this with @alber70g, we've concluded that we should try to match the format of Pact's `Command` type as closely as possible.  So that would suggest the following changes:

```
{
  "cmds": [ { "cmd": string, "sigs": [ { "pubKey":string|null, "sig": string|null }]}]
}
```

So we're suggesting the following changes:

1. `commandSigDatas` becomes `cmds` to match https://api.chainweb.com/openapi/pact.html#tag/endpoint-send/paths/~1send/post
2. `signatureList` becomes `sigs`, matching https://api.chainweb.com/openapi/pact.html#tag/model-command
3. `signature` becomes `sig` for the same reason
4. `publicKey` becomes `pubKey` and also becomes optional, which allows this API to work transparently with `Command`.

I think `CommandSigData` is still a fine name.  It is simply `Command` with the ability to optionally specify `pubKey` at the top level to make signing simpler.

Reply:
I like these requested changes.
One of my biggest qualms with the current signing API is that I can't just send a command to the wallet to be signed, the data has been flattened, and some of the names have been changed. This is also very true in the Pact-lang-api... but that's a conversation for another time.

Reply:
My thoughts on each are as follows:
Re:
1) They aren't `cmds`. They are capable of representing states that a `cmd` is not capable of representing. They also dont carry the hash like a `cmd` does. Maybe, just maybe, I could be convinced that `payloads` is a better name since it's more accurate
2) Again, it's not the same as `sigs` in `Command` though, bc it allows optional sigs, which is something the Command type does not. 
3) Sounds good
4) Changing publicKey to `pubKey` sounds good to me. Allowing it to be optional on the other hand allows for the following:
```
[ { "sig": "<sig-here>"}, {}, { "pubKey": <key>, "sig": "<sig here>" } ]
```
This doesn't feel like it gets us any closer to compat with `Command`, and removing the pubkey makes this way less ergonomic for consumers of the api.

Reply:
Ok, I see what you're saying. I would suggest something like this then:
`CommandSigData` -> `CommandSignatureRequest`
`cmd` -> `commandPayload`
`signatureList` -> `signatureRequestList` or `requiredSigList`

Don't care too much about optional on the pubKey, but the result from my above info would be this:

```json
{
  "commandSignatureRequests": [ { "commandPayload": string, "signatureRequestList": [ { "pubKey":string|null, "sig": string|null }]}]
}
```